src/stan/variational/advi.hpp

```diff
@@ -150,18 +150,143 @@ namespace stan {
       }
 
       /**
-       * Runs stochastic gradient ascent with Adagrad.
+       * Adaptively set hyperparameters for ADVI.
+       *
+       * @param variational variational distribution
+       * @param tuning_iter number of tuning iterations
+       * @return optimal value of eta in a coarse grid
+       */
+      double tune(Q& variational, int tuning_iter) const {
+        static const char* function =
+          "stan::variational::advi::tune";
+
+        // Gradient parameters
+        Q elbo_grad = Q(model_.num_params_r());
+
+        // Learning rate parameters
+        Q params_prop = Q(model_.num_params_r());
+        double tau = 1.0;
+        double pre_factor  = 0.9;
+        double post_factor = 0.1;
+        double eta_scaled;
+
+        // Sequence of eta values to try during tuning
+        std::queue<double> eta_sequence;
+        eta_sequence.push(1.00);
+        eta_sequence.push(0.50);
+        eta_sequence.push(0.10);
+        eta_sequence.push(0.05);
+        eta_sequence.push(0.01);
+
+        // Print progress
+        int eta_sequence_size = eta_sequence.size();
+        int m;
+        if (print_stream_) {
+          *print_stream_ << "Begin hyperparameter tuning." << std::endl;
+        }
+
+        // Initialize ELBO and convergence tracking variables
+        double elbo(0.0);
+        double elbo_best = -std::numeric_limits<double>::max();
+        double elbo_init = calc_ELBO(variational);
+        double eta;
+        double eta_best;
+
+        int iter_tune;
+        bool do_more_tuning = true;
+        while (do_more_tuning) {
+          // Try next eta
+          eta = eta_sequence.front();
+          eta_sequence.pop();
+
+          for (iter_tune = 1; iter_tune <= tuning_iter; ++iter_tune) {
+            m = (eta_sequence_size - eta_sequence.size() - 1) *
+              tuning_iter + iter_tune; // # of total tuning iterations
+            stan::services::variational::print_progress(
+              m, 0, tuning_iter*eta_sequence_size,
+              tuning_iter, true, "", "", *print_stream_);
+
+            // Compute gradient of ELBO
+            calc_ELBO_grad(variational, elbo_grad);
```

Comment:
question: is this the right pattern? what does `calc_ELBO_grad` actually calculate? it doesn't return anything, so I'm not sure what it does.
(no need to take any action... just want to know.)